File: myurlshortner-consumer/myurlshortner-consumer/src/main/scala/com/acme/myurlshortner/consumer/domain/entity/useragent/Browser.scala
package com.acme.myurlshortner.consumer.domain.entity.useragent

// Data Type with is-a relationship
sealed abstract class BrowserType(
  val value: String,
  val userAgentValue: String | Nothing
)

case object MozillaFirefox extends BrowserType("Mozilla Firefox", "Firefox")
case object Chrome         extends BrowserType("Google Chrome", "Chrome")    
case object Safari         extends BrowserType("Safari", "Safari")           
case object OtherBrowser   extends BrowserType("Other", null)                 

File: myurlshortner-consumer/src/main/kotlin/com/acme/myurlshortner/consumer/domain/useragent/Browser.kt
package com.acme.myurlshortner.consumer.domain.useragent

sealed class Browser {
    val userAgentValue: String?

    constructor(userAgentValue: String?) {
        this.userAgentValue = userAgentValue
    }

    data object Firefox : Browser("Firefox")
    data object Safari : Browser("Version")
    data object Chrome : Browser("Chrome")
    data object Other : Browser(null)
}
